### PR TITLE
HHH-10591 - Custom Scanner throws ClassCast when loaded from TempClas…

### PR DESCRIPTION
**Pull request by vladmihalcea**
https://hibernate.atlassian.net/browse/HHH-10591

### CODE REVIEW

**Comment by vladmihalcea** [comment]
I'm not sure about this change because if we set the jpaTempClassLOader to null then maybe we break something elsewhere.

**Comment by sebersole** [comment]
We need to get better at understanding when to use tempClassLoader and when not to.  Like here:

1. The tempClassLoader should absolutely NOT be used to load the Scanner class
2. The tempClassLoader should absolutely ideally be used to perform the work done by the Scanner.

But even in that second option we'd then have to be very careful that we do not hold any class references obtained from the tempClassLoader.

Bottom line, I am not so sure this is as simple as passing along null here.

**Comment by Janario** [comment]
It could be resolved at TempClassLoader changing the load of classes that belongs to the JPA impl (hibernate) to load from current classloader

e.g.
CustomAutoScanner > tempClassLoader
 super org.hibernate.boot.archive.scan.internal.StandardScanner > [org.hibernate] current classloader

Something like:
https://github.com/wildfly/wildfly/blob/master/jpa/subsystem/src/main/java/org/jboss/as/jpa/classloader/TempClassLoader.java#L64
```
@Override
protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
  if ( name.startsWith( "org.hibernate" ) ) {
    return Class.forName(name, resolve, delegate);
  }
  return findClass( name );
}

```

But I am not sure if this should be resolved at wildfly-jpa or hibernate. 
Also I don't know how wildfly-jpa would know the packages of JPA impl

What do you think?

**Comment by sebersole** [comment]
Yes that could work in the short term.
